### ISSUE TITLE
facade-worker.py - no SQL escaping

### ISSUE THREAD

**Issue by lukaszgryglicki** [brianwarner/facade#8]
If there is a ' or other special character in company name (I've imported rather big mappings from my other project, and some companies have `'` in their names) then affiliations fails with SQL error.
This is due to usage of code like this:
```
insert = ("INSERT INTO some_table (col1, col2) VALUES ('%s', '%s')" % (col1, col2))
cursor.execute(insert)
```
While python / SQL best practice suggests:
```
insert = ("INSERT INTO some_table (col1, col2) VALUES (%s, %s)")
cursor.execute(insert, (col1, col2))
```
Please also note that:
If there is a single item to escape, then You must use 1-tupple:
`cursor.execute(insert, (col, ))`
This will fail !:
`cursor.execute(insert, col)`

I'm fixing this in my docker fork here: https://github.com/lukaszgryglicki/facade
My fork purpose is to create docker container with preinstalled facade with our @cncf mappings.

So I will create PR fixinig this in a few hours!

I cannot assign myself - but please assigne me if You have rights to do so.

... not speaking about SQL injections ... :P

**Comment by lukaszgryglicki** [comment]
Opened PR with fix: https://github.com/brianwarner/facade/pull/9

**Comment by lukaszgryglicki** [comment]
Updated few more Python files and one PHP file.

**Comment by brianwarner** [comment]
I'm really sorry that I'm just seeing this now!  I rarely log into github, and was under the impression I was the only one using Facade right now.  Now I know better...

Thanks for this.  I'll have a look at your PR.  I ran into that same issue, and pushed a fix that could have been in parallel with yours.  However, once I go through your PR in detail I should be able to tell for sure if there was something I missed, and will merge accordingly.  It is entirely possible.

Side note:
Unfortunately when you forked it, Facade wasn't quite ready and was very unstable, so I really apologize for the headaches it probably gave you...  I've done a lot of work on it over the past few weeks and am just about to declare it ready for others to use, so you might want to have a look at the current master.  Some of the tables have changed, and I added utf-8 support, so you will want to set it up in a different database.  The setup script now creates the database and user for you, too.

**Comment by brianwarner** [comment]
Ok, I've looked through the commits.  I greatly appreciate this, and truth be told I learned a few new things.  This looks like a much better way and it beats the clunky escapes I was using to handle apostrophes.

Unfortunately there's a lot of merging to be done, but it's worth it.  I'll roll your changes through and address any other affected places in the code.

Thanks again!

**Comment by lukaszgryglicki** [comment]
Well I have a fork of Your facade running there: http://85.222.70.18:8888
I've made it work with Docker.
But I'm very busy ATM (see my github profile, I'm working on other stuff for CNCF).
Specially:
https://github.com/cncf/gitdm
and
https://github.com/cncf/velocity

My fork with Docker support is here:

My changes are mainly related to utf-8 and sql injectsions/escaping.
We're quite out of sync atm, I'll ask my boss to give me some time to sync.
If He approves I'll merge my fork with Your newest master.
https://github.com/lukaszgryglicki/facade

Best Regards.

**Comment by brianwarner** [comment]
Awesome!  That's really cool.  I would be very interested in your thoughts on utf-8 support - namely, whether I even did it correctly...  I will admit I struggled, so that would be quite welcome.

I would recommend waiting a day or two before doing anything major.  I'm trying to prepare things for an initial release, which is why so many big things changed in the past few weeks.  My hope was to push any major changes that would break backwards compatibility before anyone uses it in a serious way, though you beat me to that ;-).  I think I'm pretty close, after I finish merging your PR.  I'll drop an "all clear" in this thread when I've hit that point, and then we can look at merging the two.

**Comment by lukaszgryglicki** [comment]
Well, I'm not a Python specialist, actually all I did was a 1 day trial & error + googling.
I was just trying to make Facade work for our case.

**Comment by brianwarner** [comment]
Sounds good, I will let you know.  I am reasonably certain that the changes from your pull request are the last bit, unless something crops up in the next day or so.  I don't plan any more changes to existing database tables from this point forward, so I'm mainly just keeping an eye out for bugs.

I should also note that you ought to be able to export your various configs on the "configuration" page, update Facade, run the setup script, and then re-import everything without losing your current configs.  It'll take a while to rebuild your analysis data, but it's designed to detect missing data and backfill it.

I addressed the utf-8 issue by creating a database that uses utf8mb4, and updating the connection to use utf-8.  It's not exactly obvious, so the setup script now creates everything (if desired).

**Comment by lukaszgryglicki** [comment]
I will just merge Your master into my fork and rerun everythiong from scratch using Docker.
I'll wait few days.

**Comment by brianwarner** [comment]
Ok.  Things look good on this end after running it a few times, and I'd say you're good to go whenever you want.

I had a look at your fork and have some notes on what changed since you forked.  I'll start a new issue and drop my thoughts in there.

**Comment by lukaszgryglicki** [comment]
Great thanks, I'll work on this tomorrow probably, I have a "green light" from Dan.
I only need to take a look at some Docker issue in other project that should be fast and I'll be back on this.

**Comment by brianwarner** [comment]
So in the course of the recent work, I went back to includes/display.php and noticed the addition of $db->real_escape_string, from commit b6af571f23163465d0f04e68e39baa2d95399338.

I believe this is actually redundant, as all user input gets passed through the sanitize_input() function, which already calls real_escape_string (plus a few other things to be safe).  By the time things get to that part of the process, they should have been well and thoroughly escaped already.  If they aren't, it either means sanitize_input() wasn't doing its job right and needs to be fixed, or the POST variable wasn't going through sanitize_input() and needs to be fixed.

If it's all the same to you, I'm going to back out those changes for simplicity, and to avoid duplicate database calls.  Thoughts?

**Comment by lukaszgryglicki** [comment]
Really no idea.
I was not aware if You sanitize input anywhere else.
If it is already sanitized on server side then OK, one place is enough.
But if it is sanitized on client side then I would leave santitize on server side too.
I'm a fan of security so I think that all inputs should be double sanitized - one on client side and one on server side (to avoid possible hacks etc).
If we need to sanitize only once I would leave server side one.

Anyway - additional sanitize won't slow down and won't hurt.
It just helps to ensure code is safe and one more plus: if anybody anytime use it via famous Copy Paste's method (an ancient known Philosopher called Copy Paste) then copied code used as template will be already safe :P .

**Comment by brianwarner** [comment]
ok cool.  Yes, it's the server side, I haven't done anything client-side.
Javascript is not my forte...

sanitize_input() is in db.php and it trims input to a defined length, trims
whitespace, strips slashes, escapes html special characters, and then runs
real_escape_string.

I did it through a function so I wouldn't have to remember to do all that
stuff each time ;-)


On Thu, Jun 22, 2017 at 12:20 PM, Łukasz Gryglicki <notifications@github.com

**Comment by brianwarner** [comment]
Wait a second, you were totally right.  I was just thinking about this in terms of escaping user input, but your patch fixed a bug in the display.  Sorry about that, I've reverted my modification with an appropriately contrite commit message.
